lib/logstash/codecs/identity_map_codec.rb

```diff
@@ -0,0 +1,255 @@
+# encoding: utf-8
+require "logstash/namespace"
+# require "concurrent"
+require "thread_safe"
+
+# This class is a Codec duck type
+# Using Composition, it maps from a stream identity to
+# a cloned codec instance via the same API as a Codec
+# it implements the codec public API
+
+module LogStash module Codecs class IdentityMapCodec
+  # subclass of Exception, LS has more than limit (20000) active streams
+  class IdentityMapUpperLimitException < Exception; end
+
+  module EightyPercentWarning
+    extend self
+    def visit(imc)
+      current_size, limit = imc.current_size_and_limit
+      return if current_size < (limit * 0.8)
+      imc.logger.warn("IdentityMapCodec has reached 80% capacity",
+        :current_size => current_size, :upper_limit => limit)
+    end
+  end
+
+  module UpperLimitReached
+    extend self
+    def visit(imc)
+      current_size, limit = imc.current_size_and_limit
+      return if current_size < limit
+      # we hit the limit
+      # try to clean out stale streams
+      current_size, limit = imc.map_cleanup
+      return if current_size < limit
+      # we are still at the limit and all streams are in use
+      imc.logger.error("IdentityMapCodec has reached 100% capacity",
+          :current_size => current_size, :upper_limit => limit)
+      raise IdentityMapUpperLimitException.new
+    end
+  end
+
+  class MapCleaner
+    def initialize(imc, interval)
+      @imc, @interval = imc, interval
+      @running = false
+    end
+
+    def start
+      return self if running?
+      @running = true
+      @thread = Thread.new(@imc) do |imc|
+        loop do
+          sleep @interval
+          break if !@running
+          imc.map_cleanup
+        end
+      end
+      self
+    end
+
+    def running?
+      @running
+    end
+
+    def stop
+      return if !running?
+      @running = false
+      @thread.wakeup
+    end
+  end
+
+  # A composite class to hold both the codec and the eviction_timeout
+  # instances of this Value Object are stored in the mapping hash
+  class CodecValue
+    attr_reader :codec
+    attr_accessor :timeout
+
+    def initialize(codec)
+      @codec = codec
+    end
+  end
+
+  #maximum size of the mapping hash
+  MAX_IDENTITIES = 20_000
+
+  # time after which a stream is
+  # considered stale
+  # each time a stream is accessed
+  # it is given a new timeout
+  EVICT_TIMEOUT = 60 * 60 * 1 # 1 hour
+
+  # time that the cleaner thread sleeps for
+  # before it tries to clean out stale mappings
+  CLEANER_INTERVAL = 60 * 5 # 5 minutes
+
+  attr_reader :identity_map
+  attr_accessor :base_codec, :cleaner
+
+  def initialize(codec)
+    @base_codec = codec
+    @base_codecs = [codec]
+    @identity_map = ThreadSafe::Hash.new &method(:codec_builder)
+    @max_identities = MAX_IDENTITIES
+    @evict_timeout = EVICT_TIMEOUT
+    @cleaner = MapCleaner.new(self, CLEANER_INTERVAL)
+    @decode_block = lambda {|*| }
+  end
+
+  # a constructional method
```

Comment:
I don't mind the verbosity in the comments here, but in general I think its nicer to cut the redundancy and just say label the group of methods as 'constructional' or 'builder-like' or something